### DISCUSSION TITLE
关于C++20 module编译错误的问题

### DISCUSSION THREAD

**Discussion by Extra-Creativity** [xmake-io/xmake#4666] [Q&A] [answered]
把xmake的C++ module的internal partition测试里的hello_internal.mpp改名为hello_internal.cpp（xmake.lua也相应改下）就会分析出自己对自己的循环依赖，感觉是bug吧？用的是g++，11和13都有这问题。

**Comment by waruqi** [comment]
试下 https://github.com/xmake-io/xmake/pull/4321 这个 patch 呢，最近再重构

**Reply by Extra-Creativity** [reply]
还是有这个问题

**Reply by waruqi** [reply]
@Arthapz If we rename hello_internal.mpp to hello_internal.cpp in the internal partition test of xmake's C++ module (and xmake.lua accordingly) and it'll cause cyclic dependencies.

Any idea?

**Comment by waruqi** [comment]
再试试 modules 这个分支呢，最近对模块重构 有改进过对循环依赖的检测支持。。`xmake update -s github:xmake-io/xmake#modules`

**Reply by Extra-Creativity** [reply]
好像没解决

**Reply by waruqi** [reply, marked as answer]
这个目前没有更好的办法，你用的编译器不支持原生的模块依赖扫描，走的是 xmake 内置的扫描实现，是有局限性的，你只能改成 .mpp 。。

https://github.com/xmake-io/xmake/blob/963f544355305376c66e8fc9a98af19396c2a4a5/xmake/rules/c%2B%2B/modules/modules_support/common.lua#L548

除非你用最新的 gcc，clang 或者切到 msvc  ，让 xmake 优先使用编译器原生提供的 模块依赖扫描接口，才能提供更加精准的依赖分析。。

你可以升级你的编译器，确保 clang-scan-deps 可用，或者用最新支持 modules 的 gcc 版本，或者用 msvc preview
